sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltinFun.scala

```diff
@@ -1259,56 +1259,66 @@ private[lf] object SBuiltinFun {
-                        ) { dstContract =>
-                          // If the destination and src templates are the same, we skip the computation
-                          // of the destination template's view and the validation of the contract info.
-                          if (dstTplId == srcTplId)
-                            k(SAny(Ast.TTyCon(dstTplId), dstArg))
-                          else
-                            checkContractUpgradable(coid, srcContract, dstContract) { () =>
-                              executeExpression(machine, SEPreventCatch(dstView)) { dstViewValue =>
-                                if (srcViewValue != dstViewValue) {
-                                  Control.Error(
-                                    IE.Upgrade(
-                                      IE.Upgrade.ViewMismatch(
-                                        coid,
-                                        interfaceId,
-                                        srcTplId,
-                                        dstTplId,
-                                        srcView = srcViewValue.toUnnormalizedValue,
-                                        dstView = dstViewValue.toUnnormalizedValue,
-                                      )
-                                    )
-                                  )
-                                } else
-                                  k(SAny(Ast.TTyCon(dstTplId), dstArg))
+      viewInterfaceSafe(machine, interfaceId, srcTplId, srcArg) { mbSrcView =>
+        resolvePackageName(machine, pkgName) { pkgId =>
+          val dstTplId = srcTplId.copy(packageId = pkgId)
+          machine.ensurePackageIsLoaded(
+            dstTplId.packageId,
+            language.Reference.Template(dstTplId),
+          ) { () =>
+            ensureTemplateImplementsInterface(machine, interfaceId, coid, dstTplId) {
+              fromInterface(machine, srcTplId, srcArg, dstTplId) {
+                case None =>
+                  Control.Error(IE.WronglyTypedContract(coid, dstTplId, srcTplId))
+                case Some(dstArg) =>
+                  viewInterface(machine, interfaceId, dstTplId, dstArg) { dstView =>
+                    def withViewValue(mbSrcViewValue: Option[SValue]): Control[Question.Update] = {
+                      getContractInfo(
+                        machine,
+                        coid,
+                        dstTplId,
+                        dstArg,
+                        allowCatchingContractInfoErrors = false,
+                      ) { dstContract =>
+                        // If the destination and src templates are the same, we skip the computation
+                        // of the destination template's view and the validation of the contract info.
+                        if (dstTplId == srcTplId)
+                          k(SAny(Ast.TTyCon(dstTplId), dstArg))
+                        else
+                          checkContractUpgradable(coid, srcContract, dstContract) { () =>
```

Comment:
This looks like it comes too late.

What I'd expect to happen is
1. upgrade arg to target template
2. compute view on the target template

Reply:
Re 1: The `fromInterface` call on line 1270 does the upgrade from srcArg to the target template by calling `importValue` internally
Re 2: We then calculate the interface for dstArg/dstTmplId with `viewInterface` on line 1274

These seem to match what you're expecting to happen - the `viewInterfaceSafe` call they're wrapped in above is only used to get the mbSrcView which we compare dstView below in order to emit a warning. We won't need it if we decide to drop the warning per discussion below.